File: src/main/java/com/HOIIVUtils/clauzewitz/localization/LocalizationCollection.java
package com.HOIIVUtils.clauzewitz.localization;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public class LocalizationCollection extends HashMap<File, List<Localization>> {
    /**
     * very important for lookup by key optimization.
     */
    private final ConcurrentMap<String, Localization> localizationKeyMap = new ConcurrentHashMap<>();

    // Custom method to add a Localization
    public void add(Localization localization, File file) {
        if (localization == null || file == null) {
            throw new IllegalArgumentException("Localization and file must not be null");
        }
        List<Localization> localizationsList = this.computeIfAbsent(file, k -> new ArrayList<>());
        localizationsList.add(localization);
        localizationKeyMap.put(localization.ID(), localization); // Indexing for fast retrieval
    }

    // Custom method to remove a Localization
    public boolean remove(Localization localization) {
        if (localization == null) {
            return false;
        }
        boolean removed = this.values().parallelStream().anyMatch(localizationsList -> localizationsList.remove(localization));
        if (removed) localizationKeyMap.remove(localization.ID()); // Remove from index
        removeListIfEmpty();
        return removed;
    }

    private boolean removeListIfEmpty() {
        return this.values().removeIf(List::isEmpty);
    }

    // Custom method to get all Localizations
    public List<Localization> getAll() {
        return localizationKeyMap.values().parallelStream().toList();
    }

    public List<Localization> getAll(List<String> localizationKeys) {
        return localizationKeyMap.values().parallelStream()
                .filter(localization -> localizationKeys.contains(localization.ID()))
                .toList();
    }

    // total number of localizations
    public int numLocalization() {
        return this.values().parallelStream().mapToInt(List::size).sum();
    }

    @Override
    public List<Localization> put(File key, List<Localization> value) {
        throw new UnsupportedOperationException("Use add(Localization, File) method instead");
    }

    @Override
    public List<Localization> remove(Object key) {
        throw new UnsupportedOperationException("Use remove(Localization) method instead");
    }

    @Override
    public boolean containsKey(Object key) {
        throw new UnsupportedOperationException("Use containsLocalizationKey(String) method instead");
    }

    public boolean containsLocalizationKey(String id) {
        return localizationKeyMap.containsKey(id); // Fast lookup from index
    }

    public Localization get(String key) {
        return localizationKeyMap.get(key); // Fast lookup from index
    }

    // Method to filter localizations by status
    public List<Map.Entry<File, List<Localization>>> filterByStatus(Localization.Status status) {
        return this.entrySet().parallelStream()
                .map(entry -> Map.entry(entry.getKey(),
                        entry.getValue().parallelStream()
                                .filter(localization -> localization.status() == status)
                                .toList()))
                .filter(entry -> !entry.getValue().isEmpty())
                .toList();
    }

    public Localization replace(String key, Localization localization) {
        if (localization == null) {
            throw new IllegalArgumentException("Localization must not be null");
        }

        for (Map.Entry<File, List<Localization>> entry : this.entrySet()) {
            List<Localization> localizationsList = entry.getValue();
            for (int i = 0; i < localizationsList.size(); i++) {
                Localization currentLocalization = localizationsList.get(i);
                if (currentLocalization.ID().equals(key)) {
                    // Replace the localization and return the old one
                    localizationsList.set(i, localization);
                    return currentLocalization;
                }
            }
        }
        // Return null if no localization with the given key was found
        return null;
    }


    public File getLocalizationFile(String key) {
        return this.entrySet().parallelStream()
                .filter(entry -> entry.getValue().parallelStream()
                        .anyMatch(localization -> localization.ID().equals(key)))
                .map(Map.Entry::getKey)
                .findFirst().orElse(null);
    }
}

File: _src_old/main/scala/com/hoi4utils/hoi4mod/localization/LocalizationCollection.scala
package com.hoi4utils.hoi4mod.localization

import java.io.File
import scala.collection.concurrent.TrieMap
import scala.collection.mutable
import scala.collection.mutable.ListBuffer
import scala.util.boundary

class LocalizationCollection extends mutable.HashMap[File, ListBuffer[Localization]] {

  /**
   * very important for lookup by key optimization.
   */
  final private val localizationKeyMap = new TrieMap[String, Localization]

  // Custom method to add a Localization
  def add(localization: Localization, file: File): Unit = {
    if (localization == null || file == null) throw new IllegalArgumentException("Localization and file must not be null")

    val localizationList = this.getOrElseUpdate(file, new ListBuffer[Localization])
//    val localizationsList = this.computeIfAbsent(file, (k: File) => new ListBuffer[Localization])
    localizationList.addOne(localization)
    localizationKeyMap.put(localization.id, localization) // Indexing for fast retrieval
  }

  // Custom method to remove a Localization
  def remove(localization: Localization): Boolean = {
    var anyRemoved = false
    if (localization == null) return false
    this.foreach { case (_, localizationsList) =>
      val removed = localizationsList.exists(_.equals(localization))
      if (removed) {
        localizationsList -= localization
        localizationKeyMap.remove(localization.id) // Remove from index
        anyRemoved = true
      }
    }
    if (anyRemoved) {
      removeListIfEmpty()
    }
    anyRemoved
//    val removed = this.values.parallelStream.anyMatch((localizationsList: List[Localization]) => localizationsList.remove(localization))
//    if (removed) localizationKeyMap.remove(localization.ID) // Remove from index
//    removeListIfEmpty()
//    removed
  }

  private def removeListIfEmpty(): Unit = {
    //this.values.removeIf(List.isEmpty)
    this.filterInPlace((_, localizationsList) => localizationsList.nonEmpty)
  }

  // Custom method to get all Localizations
  def getAll: Iterable[Localization] = {
    localizationKeyMap.values
    //localizationKeyMap.values.parallelStream.toList
  }

  def getAll(localizationKeys: Iterable[String]): Iterable[Localization] = {
    localizationKeyMap.values.filter { localization => localizationKeys.toList.contains(localization.id) }
    //localizationKeyMap.values.parallelStream.filter((localization: Localization) => localizationKeys.contains(localization.ID)).toList
  }

  // total number of localizations
  def numLocalization: Int = {
    //localizationKeyMap.size
    this.values.map(_.size).sum
    //this.values.parallelStream.mapToInt(util.List.size).sum
  }

  override def put(key: File, value: ListBuffer[Localization]): Option[ListBuffer[Localization]] = {
    throw new UnsupportedOperationException("Use add(Localization, File) method instead")
  }

  override def remove(key: File): Option[ListBuffer[Localization]] = throw new UnsupportedOperationException("Use remove(Localization) method instead")

  //override def containsKey(key: AnyRef) = throw new UnsupportedOperationException("Use containsLocalizationKey(String) method instead")

  /**
   * Checks if the given localization ID is localized (has a localization entry).
   * @param id the ID of the localization to check
   * @return true if the localization is localized, false otherwise
   */
  def containsLocKey(id: String): Boolean = {
    // Fast lookup from index
    localizationKeyMap.keySet.contains(id)
    // localizationKeyMap.containsKey(id)
  }

  def get(key: String): Option[Localization] = localizationKeyMap.get(key) // Fast lookup from index

  // Method to filter localizations by status
//  def filterByStatus(status: Localization.Status): Iterable[util.Map.Entry[File, util.List[Localization]]] = {
//    this.entrySet.parallelStream.map((entry: util.Map.Entry[File, util.List[Localization]]) => util.Map.entry(entry.getKey, entry.getValue.parallelStream.filter((localization: Localization) => localization.status eq status).toList)).filter((entry: util.Map.Entry[File, util.List[Localization]]) => !entry.getValue.isEmpty).toList
//  }
  def filterByStatus(status: Localization.Status): Iterable[(File, Iterable[Localization])] = {
  // todo parallel stream?
    this.map { case (file, localizations) =>
      val filteredLocalizations = localizations.filter(_.status == status)
      if (filteredLocalizations.nonEmpty) Some(file -> filteredLocalizations) else None
    }.flatten
//      this.entrySet.asScala // Convert the Java Set to Scala Set
//      .par // Parallel processing
//      .map { entry =>
//        val filteredLocalizations = entry.getValue.asScala // Convert the Java List to Scala List
//          .filter(_.status == status) // Filter by status
//        if (filteredLocalizations.nonEmpty) Some(entry.getKey -> filteredLocalizations) else None
//      }
//      .flatten // Remove any None values
  }

  def replace(key: String, localization: Localization): Option[Localization] = boundary {
    if (localization == null) throw new IllegalArgumentException("Localization must not be null")

    foreach { case (file, localizationsList) =>
      val index = localizationsList.indexWhere(_.id == key)
      if (index != -1) {
        val prevLocalization = localizationsList(index)
        localizationsList(index) = localization
        localizationKeyMap.put(localization.id, localization) // Update index
        boundary.break(Some(prevLocalization))
      }
    }
    None
  }
//    for (entry <- this.entrySet) {
//      val localizationsList = entry.getValue
//      for (i <- 0 until localizationsList.size) {
//        val currentLocalization = localizationsList.get(i)
//        if (currentLocalization.ID == key) {
//          // Replace the localization and return the old one
//          localizationsList.set(i, localization)
//          return currentLocalization
//        }
//      }
//    }
//    // Return null if no localization with the given key was found
//    null

  def getLocalizationFile(key: String): File = {
    // todo parallel stream?
    this.filter { case (file, localizations) => localizations.exists(_.id.equals(key)) }.keys.headOption.orNull
    //this.entrySet.parallelStream.filter((entry: util.Map.Entry[File, util.List[Localization]]) => entry.getValue.parallelStream.anyMatch((localization: Localization) => localization.ID == key)).map(util.Map.Entry.getKey).findFirst.orElse(null)
  }

  def getLocalizationFiles: Iterable[File] = this.keys
}
